replica/table.cc

```diff
@@ -2805,7 +2805,53 @@ class compaction_group::table_state : public compaction::table_state {
         return _t.is_auto_compaction_disabled_by_user();
     }
     bool tombstone_gc_enabled() const noexcept override {
-        return _t._tombstone_gc_enabled;
+        /**
+         * #14870 - block our GC iff there is potential data
+         * left in commitlog that could cause resurrections
+         * on crash+replay.
+         * 
+         * We take a pessimistic approach and check for lowest
+         * rp still left in CL, and if any sstable we process 
+         * has high RP higher or equal than this, we must 
+         * block GC for now.
+         * 
+         * Note: this messes with several tests for compaction
+         * that expect a flush + compaction + check for data to
+         * have done GC already. See database.cc for workaround
+         * for this. 
+        */
+        if (!_t._tombstone_gc_enabled) {
+            return true;
+        }
+        auto cl = _t.commitlog();
+        if (!cl) {
+            return true;
+        }
+        auto lowrp = cl->lowest_active_table_rp(schema()->id());
+        if (!lowrp) {
+            return true;
+        }
+        bool ok = true;
+        auto check = [&](const shared_sstable& s) {
+            if (*lowrp <= s->get_stats_metadata().position) {
+                if (tlogger.is_enabled(log_level::debug)) {
+                    auto active = cl->active_cf_ids();
+                    tlogger.debug("GC blocked for {}:{} ({}). SSTable rp {} <= CL low mark {}",
+                        schema()->ks_name(), schema()->cf_name(), schema()->id(),
+                        s->get_stats_metadata().position, *lowrp
+                    );
+                    tlogger.debug("Active IDs (blocking): {}", active);
+                }
+
+                ok = false;
+                return stop_iteration::yes;
+            }
+            return stop_iteration::no;
+        };
+        if (main_sstable_set().for_each_sstable_until(check) == stop_iteration::no) {
+            maintenance_sstable_set().for_each_sstable_until(check);
+        }
```

Comment:
I think this might be too coarse. We should only check the actually compacted stables. Which might require this to be moved to `compaction.cc`.

Reply:
Or add an interface method to process them, or add the set as a param to this call. In compaction manager we've lost a bunch of the info required here, which is a bit of a nuisance... (database, actual table, CL etc..)

Reply:
Next problem is that this is checked from a bunch of places with various levels of info/object available. 
Maybe a compromise is to at least filter on eligible_for_compaction(sstable) here. Not sure how much/little this would filter out.

Reply:
We could add `std::vector<shared_sstable>` parameter. If the list is empty, we check the main+maint sets, as you do now. If the list is non-empty, we check only passed-in sstables.
Callers pass sstables if they have at hand, otherwise pass empty list.

Reply:
I added a patch that extends the interface and hopefully sends the correct sstables as parameters when applicable. 
Tests still pass, so there is always that. :-)

Reply:
CL tests that is. Not all compaction tests. But those fail regardless with this approach...